File: 24_Graph-II/H-TUF-Find the MST weight - Prim's Algo.java
// TUF: Find the MST weight using Prim's Algorithm
// https://takeuforward.org/plus/dsa/problems/find-the-mst-weight

/*
Problem Statement:
Given a weighted, undirected and connected graph of V vertices and E edges. 
The task is to find the sum of weights of the edges of the Minimum Spanning Tree (MST).

Approach - Prim's Algorithm:
1. Use a priority queue (min-heap) to store {weight, node}
2. Start from node 0, add it to MST and mark as visited
3. Add all adjacent edges to priority queue
4. Pick minimum weight edge that connects to an unvisited node
5. Add that node to MST and repeat until all nodes are visited
6. Keep track of total MST weight

Time Complexity: O(E log E) where E is number of edges
Space Complexity: O(V + E) for adjacency list and visited array
*/

import java.util.*;

class Solution {
    // Pair class to store {weight, node}
    static class Pair implements Comparable<Pair> {
        int weight;
        int node;
        
        Pair(int weight, int node) {
            this.weight = weight;
            this.node = node;
        }
        
        @Override
        public int compareTo(Pair other) {
            return this.weight - other.weight;
        }
    }
    
    public int spanningTree(int V, List<List<List<Integer>>> adj) {
        // Priority queue to store {weight, node}
        PriorityQueue<Pair> pq = new PriorityQueue<>();
        
        // Visited array to track nodes in MST
        boolean[] visited = new boolean[V];
        
        // Start from node 0 with weight 0
        pq.add(new Pair(0, 0));
        
        int mstWeight = 0;
        
        while (!pq.isEmpty()) {
            Pair current = pq.poll();
            int wt = current.weight;
            int node = current.node;
            
            // If already visited, skip
            if (visited[node]) {
                continue;
            }
            
            // Mark as visited and add weight to MST
            visited[node] = true;
            mstWeight += wt;
            
            // Add all adjacent edges to priority queue
            for (List<Integer> edge : adj.get(node)) {
                int adjNode = edge.get(0);
                int edgeWeight = edge.get(1);
                
                if (!visited[adjNode]) {
                    pq.add(new Pair(edgeWeight, adjNode));
                }
            }
        }
        
        return mstWeight;
    }

    public static void main(String[] args) {
        // Test Case 1
        int V1 = 3;
        List<List<List<Integer>>> adj1 = new ArrayList<>();
        
        for (int i = 0; i < V1; i++) {
            adj1.add(new ArrayList<>());
        }
        
        // Edge 0-1 with weight 5
        adj1.get(0).add(Arrays.asList(1, 5));
        adj1.get(1).add(Arrays.asList(0, 5));
        
        // Edge 1-2 with weight 3
        adj1.get(1).add(Arrays.asList(2, 3));
        adj1.get(2).add(Arrays.asList(1, 3));
        
        // Edge 0-2 with weight 1
        adj1.get(0).add(Arrays.asList(2, 1));
        adj1.get(2).add(Arrays.asList(0, 1));
        
        Solution sol = new Solution();
        System.out.println("MST Weight (Test 1): " + sol.spanningTree(V1, adj1)); // Expected: 4
        
        // Test Case 2
        int V2 = 5;
        List<List<List<Integer>>> adj2 = new ArrayList<>();
        
        for (int i = 0; i < V2; i++) {
            adj2.add(new ArrayList<>());
        }
        
        // Adding edges
        int[][] edges = {
            {0, 1, 2}, {0, 3, 6},
            {1, 2, 3}, {1, 3, 8}, {1, 4, 5},
            {2, 4, 7},
            {3, 4, 9}
        };
        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];
            adj2.get(u).add(Arrays.asList(v, w));
            adj2.get(v).add(Arrays.asList(u, w));
        }
        
        System.out.println("MST Weight (Test 2): " + sol.spanningTree(V2, adj2)); // Expected: 16
    }
}

File: JavaScript/24_Graph-II/H-TUF-Find the MST weight - Prim's Algo.js
/**
 * TUF: Minimum Spanning Tree - Prim's Algorithm
 * https://takeuforward.org/plus/dsa/problems/minimum-spanning-tree
 *
 * Greedy algorithm using Min-Heap (Priority Queue).
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O((V + E) * log V) – Using Min-Heap
 * Space Complexity | O(V + E) – Adjacency list and visited array
 */

/**
 * Creates a Min-Heap (functional approach)
 * @returns {Object} MinHeap operations
 */
function createMinHeap() {
    const heap = [];
    
    function push(val) {
        heap.push(val);
        bubbleUp(heap.length - 1);
    }
    
    function pop() {
        if (heap.length === 0) return null;
        const min = heap[0];
        const last = heap.pop();
        if (heap.length > 0) {
            heap[0] = last;
            bubbleDown(0);
        }
        return min;
    }
    
    function isEmpty() {
        return heap.length === 0;
    }
    
    function bubbleUp(idx) {
        while (idx > 0) {
            const parent = Math.floor((idx - 1) / 2);
            if (heap[parent][0] <= heap[idx][0]) break;
            [heap[parent], heap[idx]] = [heap[idx], heap[parent]];
            idx = parent;
        }
    }
    
    function bubbleDown(idx) {
        const n = heap.length;
        while (true) {
            let smallest = idx;
            const left = 2 * idx + 1;
            const right = 2 * idx + 2;
            
            if (left < n && heap[left][0] < heap[smallest][0]) {
                smallest = left;
            }
            if (right < n && heap[right][0] < heap[smallest][0]) {
                smallest = right;
            }
            if (smallest === idx) break;
            
            [heap[smallest], heap[idx]] = [heap[idx], heap[smallest]];
            idx = smallest;
        }
    }
    
    return { push, pop, isEmpty };
}

/**
 * Prim's Algorithm for Minimum Spanning Tree
 * @param {number} V - Number of vertices
 * @param {number[][][]} adj - Adjacency list where adj[u] = [[v, weight], ...]
 * @returns {number} - MST weight
 */
function primMST(V, adj) {
    const visited = new Array(V).fill(false);
    const minHeap = createMinHeap();
    
    minHeap.push([0, 0]); // [weight, node]
    let mstWeight = 0;
    
    while (!minHeap.isEmpty()) {
        const [wt, u] = minHeap.pop();
        
        if (visited[u]) continue;
        
        visited[u] = true;
        mstWeight += wt;
        
        for (const [v, edgeWt] of adj[u]) {
            if (!visited[v]) {
                minHeap.push([edgeWt, v]);
            }
        }
    }
    
    return mstWeight;
}

/**
 * Prim's Algorithm with MST edges
 */
function primMSTWithEdges(V, adj) {
    const visited = new Array(V).fill(false);
    const minHeap = createMinHeap();
    
    minHeap.push([0, 0, -1]); // [weight, node, parent]
    let mstWeight = 0;
    const mstEdges = [];
    
    while (!minHeap.isEmpty()) {
        const [wt, u, parent] = minHeap.pop();
        
        if (visited[u]) continue;
        
        visited[u] = true;
        mstWeight += wt;
        
        if (parent !== -1) {
            mstEdges.push([parent, u, wt]);
        }
        
        for (const [v, edgeWt] of adj[u]) {
            if (!visited[v]) {
                minHeap.push([edgeWt, v, u]);
            }
        }
    }
    
    return { weight: mstWeight, edges: mstEdges };
}

// Test cases
// Graph with 5 vertices
// adj[u] = [[v, weight], ...]
const adj = [
    [[1, 2], [3, 6]],
    [[0, 2], [2, 3], [3, 8], [4, 5]],
    [[1, 3], [4, 7]],
    [[0, 6], [1, 8], [4, 9]],
    [[1, 5], [2, 7], [3, 9]]
];

console.log("MST Weight (Prim's):", primMST(5, adj));

const result = primMSTWithEdges(5, adj);
console.log("MST Weight:", result.weight);
console.log("MST Edges:", result.edges);
